jenkins/agent-ctl.sh

```diff
@@ -0,0 +1,248 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Run command with no args for usage instructions.
+
+usage() {
+  echo "Usage: $(basename "${0}") [--large] <instance> [AUTO|create|delete|update|reboot|attach|detatch]"
+  exit 1
+}
+
+
+set -o nounset
+set -o errexit
+
+DOCKER_VERSION='1.9.1-0~wheezy'
+GO_VERSION='go1.6.2.linux-amd64.tar.gz'
+TIMEZONE='America/Los_Angeles'
+
+# Defaults
+DISK_SIZE='500GB'  # TODO(fejta): 300G for local-ssd
+DISK_TYPE='pd-ssd'   # TODO(fejta): local-ssd
+IMAGE='debian-7-backports'  # TODO(fejta): debian8
+LARGE=
+MACHINE_TYPE='n1-highmem-32'  # TODO(fejta): optimize to lower values
+MASTER='jenkins-master'
+SCOPES='cloud-platform,compute-rw,storage-full'  # TODO(fejta): verify
+
+if [[ -z "${1:-}" ]]; then
+  usage
+fi
+
+while true; do
+  case "${1:-}" in
+    --large)
+      LARGE=yes
+      shift
+      ;;
+    --pr)
+      echo 'Talking to PR jenkins'
+      MASTER='pull-jenkins-master'
+      shift
+      ;;
+    *)
+      break
+      ;;
+  esac
+done
+
+if [[ -z "${LARGE}" ]]; then
+  read -p "Using ${MACHINE_TYPE} for testing. Continue [Y/n]: " ans
+  if [[ ! "${ans}" =~ '^[yY]' ]]; then
+    echo 'Add --large'
+    exit 1
+  fi
+  DISK_SIZE='200GB'
+  DISK_TYPE='pd-standard'
+  MACHINE_TYPE='n1-standard-1'
+fi
+
+INSTANCE="${1}"
+CMD="${2:-auto}"
+
+auto-agent() {
+  echo "Automatically creating ${INSTANCE}..."
+  detatch-agent
+  delete-agent
+  create-agent
+  update-agent
+  reboot-agent
+  attach-agent
+}
+
+tunnel-to-master() {
+  for i in {1..10}; do
+    if sudo netstat -anp | grep 8080 > /dev/null 2>&1 ; then
+      sleep 1
+    else
+      break
+    fi
+  done
+  sudo netstat -anp | grep 8080 && echo "8080 already used" && exit 1 \
+    || echo "Tunneling to ${MASTER}..."
+  gcloud compute ssh "${MASTER}" --ssh-flag='-L8080:localhost:8080' sleep 5 &
+  for i in {1..10}; do
+    if sudo netstat -anp | grep 8080 > /dev/null 2>&1 ; then
+      break
+    fi
+    sleep 1
+  done
+  sudo netstat -anp | grep 8080 > /dev/null 2>&1 || exit 1
+}
+
+
+master-change() {
+  sudo pip install jenkinsapi
+  tunnel-to-master
+  cmd="${1}"
+  kind="${2}"
+  ini="${HOME}/jenkins-master-creds.ini"  # /user/<user>/configure && show api token
+  python "$(dirname "${0}")/attach_agent.py" "${cmd}" "${INSTANCE}" "${kind}" "${ini}" "${MASTER}"
+  echo 'Waiting for tunnel to close...'
+  wait
+}
+
+
+detatch-agent() {
+  echo "Detatching ${INSTANCE}..."
+  master-change delete ignore-kind
+}
+
+
+attach-agent() {
+  gcloud compute ssh "${INSTANCE}" -- gcloud compute instances list "--filter=name=${INSTANCE}"
+  echo "Attaching ${INSTANCE}..."
+  if [[ "${INSTANCE}" =~ light- ]]; then
```

Comment:
ah, so the instance name does have special meaning. this needs to be better noted in the usage instructions.